### PR TITLE
Ontology Improvement Tool v2

### PR DESCRIPTION
**Pull request by Aksh77**

### CODE REVIEW

File: metadata_validation_conversion/ontology_improver/tasks.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        }).reset_index()
+        df['type'] = df['type'].apply(type_count)
+        # get project-specific species from organisms index
+        species = {}
+        for project in valid_projects:
+            if project in df['projects'].values:
+                species[get_species_for_project(project)] = project
+        df['species'] = species
+        # get existing summary statistics
+        url = f"http://backend-svc:8000/data/summary_ontologies/_search/?size=10"
```
Comment (Alexey-ebi) [inline metadata_validation_conversion/ontology_improver/tasks.py:77]: would be great to read URL value from global constants

File: metadata_validation_conversion/ontology_improver/tasks.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        res = requests.get(url)
```
Comment (Alexey-ebi) [inline metadata_validation_conversion/ontology_improver/tasks.py:78]: can be converted to:
requests.get(url).json()

File: metadata_validation_conversion/ontology_improver/utils.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+def hasAttribute(res, att):
+    if att in res and res[att] is not None:
+        if type(res[att]) is list:
+            if len(res[att]):
+                return True
+            else:
+                return False
```
Comment (Alexey-ebi) [inline metadata_validation_conversion/ontology_improver/utils.py:7]: can be converted to:
return True if len(res[att]) else False

File: metadata_validation_conversion/ontology_improver/views.py [hunk 4]
```diff
@@ -127,83 +68,85 @@ def authentication(request):
...
-    return False
+    ontology = data['ontology']
+    status_activity = ontology['status_activity']
+    status_activity.append({
+        'project': data['project'],
+        'status': data['status'],
+        'timestamp': datetime.now(tz=timezone.utc),
+        'user': data['user']
+    })
+    url = f"http://backend-svc:8000/data/ontologies/{ontology['key']}"
```
Comment (Alexey-ebi) [inline metadata_validation_conversion/ontology_improver/views.py:83]: can we read this from global constants

File: metadata_validation_conversion/ontology_improver/views.py [hunk 5]
```diff
@@ -127,83 +68,85 @@ def authentication(request):
...
+                'project': data['project'],
+                'status': 'Awaiting Assessment',
+                'timestamp': datetime.now(tz=timezone.utc),
+                'user': data['user']
+            }]
+        }
+        es.index(index='ontologies', id=new_ontology['key'], body=new_ontology)
+    # update summary stats - increment validated_count
+    for project in data['project']:
+        url = f"http://backend-svc:8000/data/summary_ontologies/{project}"
```
Comment (Alexey-ebi) [inline metadata_validation_conversion/ontology_improver/views.py:124]: read this from global constants

File: metadata_validation_conversion/ontology_improver/views.py [hunk 6]
```diff
@@ -127,83 +68,85 @@ def authentication(request):
...
 
 @csrf_exempt
 def get_ontology_details(request, id):
     if request.method != 'GET':
         return HttpResponse("This method is not allowed!\n")
     response = {}
     # get ontology type, status etc from FAANG Ontology Database
-    faang_data = model_to_dict(Ontologies.objects.get(pk=id))
-    del faang_data['verified_by_users']
+    url = f"http://backend-svc:8000/data/ontologies/{id}"
```
Comment (Alexey-ebi) [inline metadata_validation_conversion/ontology_improver/views.py:140]: read this from global constants

File: metadata_validation_conversion/ontology_improver/views.py [hunk 7]
```diff
@@ -227,73 +170,55 @@ def get_ontology_details(request, id):
...
-            obj.save()
-        except (Summary.DoesNotExist, KeyError):
-            obj = Summary.objects.create(
-                project=row['project'], \
-                species=row['species'], \
-                ontology_type_count=row['ontology_type_count'], \
-                status_count=row['status_count'])
+    es = Elasticsearch([settings.NODE], connection_class=RequestsHttpConnection, \
+                       http_auth=(settings.ES_USER, settings.ES_PASSWORD), \
+                        use_ssl=True, verify_certs=False)
```
Comment (Alexey-ebi) [inline metadata_validation_conversion/ontology_improver/views.py:185]: try to use verify_certs = True

**Review by Alexey-ebi — CHANGES_REQUESTED** [review]

File: metadata_validation_conversion/ontology_improver/tasks.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
+import json
+import requests
+import pandas as pd
+from collections import Counter
+from django.conf import settings
+from elasticsearch import Elasticsearch, RequestsHttpConnection
+from metadata_validation_conversion.celery import app
+
+es = Elasticsearch([settings.NODE], connection_class=RequestsHttpConnection, \
```
[the remark this answers is not in the document]
Comment (Aksh77) [reply metadata_validation_conversion/ontology_improver/tasks.py:9]: removed it

File: metadata_validation_conversion/ontology_improver/tasks.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        df = df.loc[df['projects'].isin(valid_projects)]
+        df = df.groupby(['projects']).agg({
+            'type': comma_separated_combine,
+            'term': 'count'
+        }).reset_index()
+        df['type'] = df['type'].apply(type_count)
+        # get project-specific species from organisms index
+        species = {}
+        for project in valid_projects:
+            if project in df['projects'].values:
```
[the remark this answers is not in the document]
Comment (Aksh77) [reply metadata_validation_conversion/ontology_improver/tasks.py:73]: .values seems to be right as per documentation https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.values.html

File: metadata_validation_conversion/ontology_improver/tasks.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+@app.task
+def update_ontology_summary():
+    r = requests.get(f'http://backend-svc:8000/data/ontologies/_search/?size=10000')
+    if r.status_code == 200:
+        ontologies = json.loads(r.content)['hits']['hits']
+        ontologies = map(lambda ontology : ontology['_source'], ontologies)
+        df = pd.DataFrame.from_dict(ontologies)[['projects', 'type', 'term']]
+        df['type'] = [', '.join(map(str, l)) for l in df['type']]
+        df = df.explode('projects')
+        valid_projects = [
```
[the remark this answers is not in the document]
Comment (Aksh77) [reply metadata_validation_conversion/ontology_improver/tasks.py]: moved to PROJECTS constant in metadata_validation_conversion/metadata_validation_conversion/constants.py

File: metadata_validation_conversion/ontology_improver/tasks.py (hunk 1 above)
Comment (Aksh77) [reply metadata_validation_conversion/ontology_improver/tasks.py]: moved to BE_SVC constant in metadata_validation_conversion/metadata_validation_conversion/constants.py

File: metadata_validation_conversion/ontology_improver/views.py (hunk 6 above)
Comment (Aksh77) [reply metadata_validation_conversion/ontology_improver/views.py]: moved to BE_SVC constant in metadata_validation_conversion/metadata_validation_conversion/constants.py

File: metadata_validation_conversion/ontology_improver/views.py [hunk 11]
```diff
@@ -127,83 +68,85 @@ def authentication(request):
...
+    res = requests.get(url)
+    if res.status_code != 200 or len(json.loads(res.content)['hits']['hits']) == 0:
+        return HttpResponse(status=404)
+    faang_data = json.loads(res.content)['hits']['hits'][0]['_source']
     response['faang_data'] = faang_data
     # parse ontology details from EBI OLS
     res = requests.get(
         "http://www.ebi.ac.uk/ols/api/terms?short_form={}".format(
-            response['faang_data']['ontology_id'])).json()
+            response['faang_data']['id'])).json()
```
[the remark this answers is not in the document]
Comment (Aksh77) [reply metadata_validation_conversion/ontology_improver/views.py]: 1. moved to OLS_API constant in metadata_validation_conversion/metadata_validation_conversion/constants.py
2. changed to ols4 and tested

File: metadata_validation_conversion/ontology_improver/views.py [hunk 12]
```diff
@@ -227,73 +170,55 @@ def get_ontology_details(request, id):
...
-            obj = Summary.objects.create(
-                project=row['project'], \
-                species=row['species'], \
-                ontology_type_count=row['ontology_type_count'], \
-                status_count=row['status_count'])
+    es = Elasticsearch([settings.NODE], connection_class=RequestsHttpConnection, \
+                       http_auth=(settings.ES_USER, settings.ES_PASSWORD), \
+                        use_ssl=True, verify_certs=False)
+    for ontology in ontologies:
+        url = f"http://backend-svc:8000/data/ontologies/{ontology['key']}"
```
[the remark this answers is not in the document]
Comment (Aksh77) [reply metadata_validation_conversion/ontology_improver/views.py]: moved to BE_SVC constant in metadata_validation_conversion/metadata_validation_conversion/constants.py

**Pull request merged**